server/src/main/java/org/elasticsearch/action/ingest/SimulatePipelineResponse.java

```diff
@@ -19,22 +19,87 @@
 
 package org.elasticsearch.action.ingest;
 
+import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
 import org.elasticsearch.common.xcontent.ToXContentObject;
 import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentParser.Token;
 
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.XContentParserUtils.ensureExpectedToken;
+
 public class SimulatePipelineResponse extends ActionResponse implements ToXContentObject {
     private String pipelineId;
     private boolean verbose;
     private List<SimulateDocumentResult> results;
 
+    @SuppressWarnings("unchecked")
+    public static final ConstructingObjectParser<SimulatePipelineResponse, Void> PARSER =
+        new ConstructingObjectParser<>(
+            "simulate_pipeline_response",
+            true,
+            a -> {
+                List<SimulateDocumentResult> results = (List<SimulateDocumentResult>)a[0];
+                boolean verbose = false;
+                if (results.size() > 0) {
+                    if (results.get(0) instanceof SimulateDocumentVerboseResult) {
+                        verbose = true;
+                    }
+                }
+                return new SimulatePipelineResponse(null, verbose, results);
+            }
+        );
+    static {
+        PARSER.declareObjectArray(
+            constructorArg(),
+            (parser, context) -> {
+                ensureExpectedToken(Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+                SimulateDocumentResult result = null;
+                while (parser.nextToken() != Token.END_OBJECT) {
+                    ensureExpectedToken(parser.currentToken(), Token.FIELD_NAME, parser::getTokenLocation);
+                    String fieldName = parser.currentName();
+                    parser.nextToken();
+                    switch(fieldName) {
```

Comment:
I would find this even more accurate if we looked at the token type upfront rather than checking first the field name and then asserting on the token type.

Reply:
But they are different for different cases?

Reply:
the subtle difference is that if something with same name but wrong type comes up, now it's rejected, while checking first the token type it would be ignored. It also makes parsing code more readable because you know exactly for which type of section you do what. At least that's what I got used to.